src/Servers/Kestrel/Core/src/Internal/Http2/Http2Connection.cs

```diff
@@ -1259,10 +1259,116 @@ public void OnHeader(ReadOnlySpan<byte> name, ReadOnlySpan<byte> value)
             }
         }
 
+        public void OnStaticIndexedHeader(int index)
+        {
+            Debug.Assert(index <= H2StaticTable.Count);
+
+            ref readonly var entry = ref H2StaticTable.Get(index - 1);
+            OnStaticIndexedHeaderCore(index, entry.Name, entry.Value);
+        }
+
+        public void OnStaticIndexedHeader(int index, ReadOnlySpan<byte> value)
+        {
+            Debug.Assert(index <= H2StaticTable.Count);
+
+            OnStaticIndexedHeaderCore(index, H2StaticTable.Get(index - 1).Name, value);
+        }
+
+        // We can't throw a Http2StreamErrorException here, it interrupts the header decompression state and may corrupt subsequent header frames on other streams.
+        // For now these either need to be connection errors or BadRequests. If we want to downgrade any of them to stream errors later then we need to
+        // rework the flow so that the remaining headers are drained and the decompression state is maintained.
+        public void OnStaticIndexedHeaderCore(int index, ReadOnlySpan<byte> name, ReadOnlySpan<byte> value)
```

Comment:
Can this be unified with OnHeader? The differences are very limited. Maybe use a nullable int for the non-static flow?